### PR TITLE
feat: LT-1042 - Ultimate end users 

### PR DESCRIPTION
**Pull request by jfaracik**
* Added Ultimate End Users endpoint
* Added `/users/me/` endpoint - Returns info about the currently signed in user
* Added a new `get` wrapper inside TestClient - this returns the json and status code from the regular `get` (simpler)
* Added tests
* Fixed sites not updating

### CODE REVIEW

File: applications/tests/tests_ultimate_end_users.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+    def test_submit_draft_with_no_ultimate_end_users_unsuccessful(self):
+        """
+        This should be unsuccessful as an ultimate end user is required when
+        there is a part which is to be incorporated into another good
+        """
+        data = {'id': self.draft.id}
+        response = self.client.post(self.url, data, **self.exporter_headers)
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
```
Comment (andytaylorkainos) [inline applications/tests/tests_ultimate_end_users.py]: can you please assert the error message returned

File: drafts/tests/tests_set_ultimate_end_users.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        data = {
+            'name': 'UK Government',
+            'address': 'Westminster, London SW1A 0AA',
+            'country': 'GB',
+            'website': 'https://www.gov.uk'
+        }
+
+        response = self.client.post(self.url, data, **self.exporter_headers)
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
```
Comment (andytaylorkainos) [inline drafts/tests/tests_set_ultimate_end_users.py]: again, please assert the response message

File: organisations/libraries/get_organisation.py [hunk 3]
```diff
@@ -1,16 +1,18 @@
...
 from organisations.models import Organisation
+from users.models import ExporterUser
 
 
 def get_organisation_by_pk(pk):
     try:
         return Organisation.objects.get(pk=pk)
     except Organisation.DoesNotExist:
-        raise Http404
+        raise NotFoundError({'organisation': 'Organisation not found - ' + str(pk)})
```
Comment (andytaylorkainos) [inline organisations/libraries/get_organisation.py]: 👍

File: organisations/views/views_site.py [hunk 4]
```diff
@@ -130,24 +130,23 @@ def get(self, request, pk):
...
                                               partial=True)
             if serializer.is_valid():
                 serializer.save()
+
                 reversion.set_user(request.user)
                 reversion.set_comment("Created Site Revision")
 
-                return JsonResponse(data={'site': serializer.data},
-                                    status=status.HTTP_200_OK)
+                return JsonResponse(data={'site': serializer.data})
```
Comment (andytaylorkainos) [inline organisations/views/views_site.py]: Any reason why there's no response status?

File: static/countries/helpers.py [hunk 5]
```diff
@@ -1,10 +1,9 @@
...
+from conf.exceptions import NotFoundError
 from static.countries.models import Country
 
 
 def get_country(pk):
     try:
         return Country.objects.get(pk=pk)
     except Country.DoesNotExist:
-        raise Http404
+        raise NotFoundError({'country': 'Country not found - ' + str(pk)})
```
Comment (andytaylorkainos) [inline static/countries/helpers.py]: 👍

File: users/views.py [hunk 6]
```diff
@@ -149,3 +150,13 @@ def get_queryset(self):
...
+
+
+class UserMeDetail(APIView):
+    authentication_classes = (ExporterAuthentication,)
+    """
+    Get the user from request
+    """
+    def get(self, request):
+        serializer = ExporterUserViewSerializer(request.user)
+        return JsonResponse(data={'user': serializer.data})
```
Comment (andytaylorkainos) [inline users/views.py]: again, no response status?

File: addresses/helpers.py [hunk 7]
```diff
@@ -0,0 +1,10 @@
+from django.http import Http404
+
+from addresses.models import Address
+
+
+def get_address(pk):
+    try:
+        return Address.objects.get(pk=pk)
+    except Address.DoesNotExist:
+        raise Http404
```
Comment (andytaylorkainos) [inline addresses/helpers.py]: Can we insert a custom `NotFoundError` here too please? - the others are so nice, why stop there? :D

File: addresses/serializers.py [hunk 8]
```diff
@@ -9,10 +9,11 @@ class AddressSerializer(serializers.ModelSerializer):
     """
     Used for serializing addresses
     """
-    address_line_1 = serializers.CharField()
-    postcode = serializers.CharField(max_length=10)
-    city = serializers.CharField()
-    region = serializers.CharField()
+    address_line_1 = serializers.CharField(error_messages={'blank': 'Enter a real building and street name'})
```
[the remark this answers is not in the document]
Comment (jfaracik) [reply addresses/serializers.py]: I was basing it off of https://design-system.service.gov.uk/patterns/addresses/#error-messages

[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply addresses/serializers.py]: Ah! then ignore me :D

File: organisations/views/views_site.py (hunk 4 above)
Comment (jfaracik) [reply organisations/views/views_site.py]: It defaults to 200

File: addresses/helpers.py (hunk 7 above)
Comment (jfaracik) [reply addresses/helpers.py]: Will add

File: users/views.py (hunk 6 above)
Comment (jfaracik) [reply users/views.py]: It defaults to 200

**Comment by efahy** [comment]
image

Multiple end users not displaying correctly for gov users viewing a case

File: applications/serializers.py [hunk 9]
```diff
@@ -83,6 +86,11 @@ def get_destinations(self, obj):
             serializer = CountrySerializer(countries, many=True)
             return {'type': 'countries', 'data': serializer.data}
 
+    def get_ultimate_end_users(self, application):
+        ultimate_end_users = get_ultimate_end_user_ids(application)
```
Comment (efahy) [inline applications/serializers.py]: Remove `_ids` - you're returning whole objects

File: applications/urls.py [hunk 10]
```diff
@@ -8,4 +8,5 @@
     path('', views.ApplicationList.as_view(), name='applications'),
     path('<uuid:pk>/', views.ApplicationDetail.as_view(), name='application'),
     path('<uuid:pk>/user/', views.ApplicationDetailUser.as_view(), name='application-pk-user'),
+    path('<uuid:pk>/', views.ApplicationUltimateEndUsers.as_view(), name='application_ultimate_end_users')
```
Comment (efahy) [inline applications/urls.py]: Why is the path here not ultimate end user specific?

File: applications/views.py [hunk 11]
```diff
@@ -34,8 +36,9 @@ def get(self, request):
 
         applications = Application.objects.filter(organisation=organisation).order_by('created_at')
         serializer = ApplicationBaseSerializer(applications, many=True)
+
         return JsonResponse(data={'applications': serializer.data},
-                            )
+                            status=status.HTTP_200_OK)
```
Comment (efahy) [inline applications/views.py]: Jan says this shouldn't be on here

File: applications/views.py [hunk 12]
```diff
@@ -153,3 +156,25 @@ def get(self, request, pk):
         )
 
         return super(ApplicationDetailUser, self).get(request, pk)
+
+
+class ApplicationUltimateEndUsers(APIView):
```
Comment (efahy) [inline applications/views.py]: This can go I believe

File: drafts/views/draft_end_user_views.py [hunk 13]
```diff
@@ -42,3 +44,88 @@ def post(self, request, pk):
...
+        """
+        Get ultimate end users associated with a draft
+        """
+        draft = get_draft(pk)
+        ultimate_end_users = get_ultimate_end_user_ids(draft)
+
+        serializer = EndUserSerializer(ultimate_end_users, many=True)
+
+        return JsonResponse(data={'ultimate_end_users': serializer.data},
+                            status=status.HTTP_200_OK)
```
Comment (efahy) [inline drafts/views/draft_end_user_views.py]: No need for this

File: applications/creators.py [hunk 14]
```diff
@@ -19,11 +19,25 @@ def create_standard_licence(draft, application, errors):
     if not GoodOnDraft.objects.filter(draft=draft):
         errors['goods'] = get_string('applications.standard.no_goods_set')
 
+    ultimate_end_user_required = False
+    for good_on_draft in GoodOnDraft.objects.filter(draft=draft):
+        if not good_on_draft.good.is_good_end_product:
+            ultimate_end_user_required = True
```
Comment (efahy) [inline applications/creators.py]: You could consider using python filters to do this also, see https://stackoverflow.com/questions/29563153/python-filter-function-single-result

File: applications/creators.py [hunk 15]
```diff
@@ -19,11 +19,25 @@ def create_standard_licence(draft, application, errors):
...
+
+    if ultimate_end_user_required:
+        if len(draft.ultimate_end_users.values_list()) == 0:
+            errors['ultimate_end_users'] = get_string('applications.standard.no_ultimate_end_users_set')
+
+        for ultimate_end_user in draft.ultimate_end_users.values_list('id', flat=True):
```
Comment (efahy) [inline applications/creators.py]: Put this in an else branch in the above `if`

File: applications/creators.py [hunk 16]
```diff
@@ -19,11 +19,25 @@ def create_standard_licence(draft, application, errors):
...
+            if not 'end_user' in errors and str(ultimate_end_user) == str(draft.end_user.id):
```
Comment (efahy) [inline applications/creators.py]: Could do with a comment explaining your thinking

File: applications/creators.py [hunk 17]
```diff
@@ -19,11 +19,25 @@ def create_standard_licence(draft, application, errors):
...
+                errors['ultimate_end_users'] = get_string('applications.standard.matching_end_user_and_ultimate_end_user')
```
Comment (efahy) [inline applications/creators.py]: Another comment please

**Pull request merged**